backend/src/main/kotlin/demo/acme/service/DataFetchService.kt

```diff
@@ -0,0 +1,112 @@
+import org.apache.commons.csv.CSVFormat
+import org.apache.commons.csv.CSVParser
+import org.slf4j.LoggerFactory
+import org.springframework.beans.factory.annotation.Value
+import org.springframework.stereotype.Service
+import java.io.BufferedReader
+import java.net.HttpURLConnection
+import java.net.URL
+import kotlin.system.measureTimeMillis
+
+
+@Service
+class DataFetchService {
+    @Value("\${datafetchservice.sourceApiUrl:}")
+    lateinit var sourceApiUrl: String
+    @Value("\${datafetchservice.storePagesEndpoint:}")
+    lateinit var storePagesEndpoint: String
+    @Value("\${datafetchservice.storeSeasonEndpoint:}")
+    lateinit var storeSeasonEndpoint: String
+    @Value("\${datafetchservice.extraDataCsv:}")
+    lateinit var extraDataCsv: String
+    @Value("\${datafetchservice.apiKey:}")
+    lateinit var apiKey: String
+    @Value("\${datafetchservice.maximumNumberOfFetchedPages:}")
+    lateinit var maximumNumberOfFetchedPages:String
+
+    private val logger = LoggerFactory.getLogger(this.javaClass)
+
+    fun fetchStoresPage(page: Int): List<Store>? {
+        return fetchData("${sourceApiUrl}${storePagesEndpoint}?page=$page") {
+            Klaxon().parseArray( it )
+        }
+    }
+
+    fun fetchStoresAndSeasons(): List<StoreAndSeason>? {
+        var result: List<StoreAndSeason>?
+        val elapsed = measureTimeMillis {
+            result = fetchData("${sourceApiUrl}${storeSeasonEndpoint}") { Klaxon().parseArray(it) }
+        }
+        logger.info("Fetched ${result?.size} store season entries in $elapsed milliseconds.")
+        return result
+    }
+
+    fun fetchCSV(): Map<Long, CsvEntry>? {
+        var result : Map<Long, CsvEntry>?
+        val elapsed = measureTimeMillis {
+            result = fetchData("${sourceApiUrl}${extraDataCsv}") {
+                val associate = CSVParser(it, CSVFormat.DEFAULT.withFirstRecordAsHeader().withTrim())
+                    .associate { csvRecord ->
+                        (csvRecord.get("Store id").toLongOrNull() ?: 0) to CsvEntry()
+                            .apply {
+                                specialField1 = csvRecord.get("Special field 1")
+                                specialField2 = csvRecord.get("Special field 2")
+                            }
+                    }
+                associate
+            }
+        }
+        logger.info("Fetched ${result?.keys?.size} csv entries in $elapsed milliseconds.")
+        return result
+    }
+
+    fun fetchStores(): Map<Long,Store>{
+        val stores = mutableMapOf<Long, Store>()
+        val elapsed = measureTimeMillis {
+            var page = 0
+            while (page < maximumNumberOfFetchedPages.toInt()) {
+                val asyncFetches = mutableListOf<Deferred<List<Store>?>>()
+                var storesBatch: List<List<Store>>
+                runBlocking {
+                    repeat(10) {
+                        asyncFetches.add(async { fetchStoresPage(page++) })
```

Comment:
What happens if one of these requests returns an error (eg. HTTP 500)?

Reply:
Well, tbh I played around with the API a little on Postman and made a false assumption. I thought Internal server errors were thrown on specific pages. But I failed to see that retrying those same pages would eventually result in a unique set of entries and that the 500 was random. I changed the data fetch method to retry requests at most 5 times before giving up whenever it gets something other than 200. 
Is it a common practice to re-attempt requests on server errors such as this? I mean, a timeout (504) or service unavailable(503) suggest as much but 500 feels vague. 
Or was the point here just to understand the API and get the data all the same? Which I totally missed :D